File: 2024/Golang/day3/part1.go
package day3

import (
	"Golang/utils"
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func Part1() {
	sum := 0
	file, err := os.Open("../Input/day3.txt")
	utils.Check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	pattern1 := regexp.MustCompile(`(mul\(\d{1,3},\d{1,3}\))|.`)
	pattern2 := regexp.MustCompile(`(\(\d{1,3},\d{1,3}\))|.`)

	for scanner.Scan() {
		line := scanner.Text()
		line = pattern1.ReplaceAllStringFunc(line, func(match string) string {
			if pattern1.MatchString(match) {
				subMatches := pattern1.FindStringSubmatch(match)
				if subMatches[1] != "" {
					return subMatches[1]
				}
			}
			return ""
		})

		line = pattern2.ReplaceAllStringFunc(line, func(match string) string {
			if pattern2.MatchString(match) {
				subMatches := pattern2.FindStringSubmatch(match)
				if subMatches[1] != "" {
					return subMatches[1] 
				}
			}
			return "" 
		})

		i := 0;
		for j := 0; j < len(line); j++ {
			if line[j] == ')' {
				firstNumber, _:= strconv.Atoi(strings.Split(line[i+1:j], ",")[0])
				secondNumber, _:= strconv.Atoi(strings.Split(line[i+1:j], ",")[1])
				sum += firstNumber * secondNumber
				i = j + 1
			}
		}
	}
	fmt.Println(sum)
}

File: 2024/Rust/day3/part1.rs
use crate::utils;
use regex::Regex;
use std::str::FromStr;

pub fn run_part1() {
    let mut sum = 0;
    let re1 = Regex::new(r"(mul\(\d{1,3},\d{1,3}\))|.").unwrap();
    let re2 = Regex::new(r"(\(\d{1,3},\d{1,3}\))|.").unwrap();
    for line in utils::read_lines("../Input/day3.txt") {
        let replaced = re1.replace_all(&line, |caps: &regex::Captures| {
            if let Some(matched) = caps.get(1) {
                matched.as_str().to_string()
            } else {
                "".to_string()
            }
        });
        let replaced_again = re2.replace_all(&replaced, |caps: &regex::Captures| {
            if let Some(matched) = caps.get(1) {
                matched.as_str().to_string()
            } else {
                "".to_string()
            }
        });
        let mut i = 0;
        let chars: Vec<char> = replaced_again.chars().collect();

        for j in 0..chars.len() {
            if chars[j] == ')' {
                if let Some(sub_string) = replaced_again.get(i + 1..j) {
                    let parts: Vec<&str> = sub_string.split(',').collect();
                    if parts.len() == 2 {
                        if let (Ok(first_number), Ok(second_number)) =
                            (i32::from_str(parts[0]), i32::from_str(parts[1])) {
                            sum += first_number * second_number;
                        }
                    }
                }
                i = j + 1;
            }
        }
    }
    println!("{}", sum);
}